ghc, haskell-infra: Improve cross-compilation

angerman opened the pull request:
###### Motivation for this change

This adds some minor changes to the GHC expressions, and tries to align the head and 8.4.2 expression. Ideally we'd have one common one among all ghc expressions I think.

This should allow building quick flavors, and building cross compilers to windows.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

angerman committed:
haskell generic builder: Use setup package database for setup-depends

Adapted from https://github.com/obsidiansystems/nixpkgs/commit/b69f420121120433220c568e4b35ade539ef60f2 by @Ericson2314

angerman committed:
ghc: paxmark all unwraped executables across the board

Shell glob works even as the exact set of executable (filenames) varries
beween configuations.

angerman committed:
haskell generic builder: Disable static libs on Windows because no -staticlib

The reason why this does not work is not that we can't built static
objects, we can, but we can't use `-staticlib` on GHC on windows.
`-staticlib` rolls all dependencies into a combined archive. While this
would work on windows if we used gnu ar and MRI script, GHC can't rely
on GNU ar, and as such has a quick archive concatenation module for GNU
and BSD archives only.

angerman committed:
haskell generic builder: Do the `links` dance only if shared is enabled.

angerman committed:
haskell infra: Adds buildFlags logic

angerman committed:
used stdenv.targetPlatform.isDarwin and not stdenv.isDarwin.

angerman committed:
ghc 8.4.2, head: Drop `libiconv` on windows.

nixpkgs#37012 and  nixpkgs#37707 introduces the setup-hooks for libiconv, which inject `-liconv` into the `NIX_LDFLAGS`. This breaks horribly on windows where the linker end up having no idea how to linke `-liconv`. The configure.ac file specifically ignores libiconv on windows.

peti commented on pkgs/development/compilers/ghc/head.nix:
```diff
@@ -15,16 +15,26 @@
+  enableTerminfo ? !targetPlatform.isWindows
 
 , version ? "8.5.20180118"
+, ghcRevision ? "e1d4140be4d2a1508015093b69e1ef53516e1eb6"
```
Is it really necessary to parameterize all aspects of the `src` attribute as function arguments? I would assume that few people actually mess with these settings, and those who want to can easily override the `src` attribute as a whole.

angerman replied:
Right! that should better be handed like this once we have the generic expression. Will drop.

peti commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -19,9 +19,11 @@ in
 , description ? ""
 , doCheck ? !isCross && (stdenv.lib.versionOlder "7.4" ghc.version)
 , doBenchmark ? false
+, doVerbose ? false
```
In the presence of `buildFlags` this attribute seems redundant?

angerman replied:
I agree that we can drop this in favor or passing `-v3` to the `buildFlags`.

peti commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -142,7 +146,9 @@ let
-    (optionalString (versionOlder "8.4" ghc.version) (enableFeature enableStaticLibraries "static"))
+    # --enable-static does not work on windows. This is a bug in GHC.
+    # --enable-static will pass -staticlib to ghc, which only works for mach-o and elf.
+    (optionalString (!hostPlatform.isWindows && versionOlder "8.4" ghc.version) (enableFeature enableStaticLibraries "static"))
```
If someone requests static builds for a given platform, then it feels like poor style to just ignore the setting and to silently build shared libraries instead. I'd prefer to see this particular limitation expressed as an assertion instead.

angerman replied:
I'll amend the PR, to 
```
enableStaticLibraries ? !hostPlatform.isWindows
```
and a
```
assert (enableStaticLibraries && hostPlatform.isWindows) "can't use `--enable-static` on windows."
```

The reason why this does not work is not that we can't built static objects, we can, but we can't use
`-staticlib` on ghc on windows. `-staticlib` rolls all dependencies into a combined archive. While this
would work on windows if we used gnu ar and MRI script, GHC can't rely on gnu ar, and as such has
a quck archive concatenation module for gnu and bsd archives only.

Ericson2314 replied:
On Nix we do indeed provide GNU ar always, so it would be nice to optionally configure GHC to expect GNU binutils just as one can configure GCC for the same reasons.

peti commented on pkgs/development/compilers/ghc/8.4.2.nix:
```diff
@@ -24,7 +24,12 @@
+  enableTerminfo ? !targetPlatform.isWindows
+
 , version ? "8.4.2"
+, ghcFlavour ? ""
```
Can you please add a small comment here explaining that `""` means "use ghc's default choice"?

angerman replied:
will do.

angerman commented:
@peti, thanks for the quick review, I'll make some amendments later today.

Ericson2314 commented on pkgs/development/compilers/ghc/8.4.2.nix:
```diff
@@ -24,7 +24,15 @@
+
+  # What flavour to build. An empty string indicates no
+  # specific flavour and falls back to ghc default values.
+, ghcFlavour ? ""
```
`, ghcFlavour = stdenv.lib.optionalString (targetPlatform != hostPlatform) "perf-cross" `

Ericson2314 replied:
Can we also backport this change to the earlier GHCs for consistency?

angerman replied:
We can. But I'd argue against it. I'd rather see us align the 8.{4,6}.y and head into a `common` form, and then graft previous version onto the `common.nix`. Otherwise we are going to end up backporting this stuff every time again across all expressions.

Ericson2314 commented on pkgs/development/compilers/ghc/8.4.2.nix:
```diff
@@ -38,11 +46,14 @@ let
     "${targetPlatform.config}-";
 
   buildMK = ''
+    BuildFlavour = ${if targetPlatform != hostPlatform then ghcCrossFlavour else ghcFlavour}
```
Simple `${ghcFlavour}` per the above

Ericson2314 commented on pkgs/development/compilers/ghc/8.4.2.nix:
```diff
@@ -165,7 +176,11 @@ stdenv.mkDerivation rec {
   # treat that as a unary `{x,y,z,..}` repetition.
   postInstall = ''
-    paxmark m $out/lib/${name}/bin/${if targetPlatform != hostPlatform then "ghc" else "{ghc,haddock}"}
+    for f in $out/lib/${name}/bin/${if targetPlatform != hostPlatform then "ghc" else "{ghc,haddock}"}; do
```
Can we just do `$out/lib/${name}/bin/*` then?

Ericson2314 commented on pkgs/development/compilers/ghc/8.4.2.nix:
```diff
@@ -38,11 +46,14 @@ let
 
   buildMK = ''
+    BuildFlavour = ${if targetPlatform != hostPlatform then ghcCrossFlavour else ghcFlavour}
+    ifneq \"\$(BuildFlavour)\" \"\"
```
Nix doesn't require `"` to be escaped here, to be clear. Does make?

Ericson2314 commented on pkgs/development/compilers/ghc/8.4.2.nix:
```diff
@@ -24,7 +24,15 @@
   # platform). Static libs are always built.
   enableShared ? !targetPlatform.useAndroidPrebuilt
```
You wish to allow dynamic linking here but not 8.2?

angerman replied:
I don't follow? That line wasn't even changed?

angerman replied:
Ahh... ok. fixed.

Ericson2314 commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -68,6 +69,10 @@ in
 
+# --enable-static does not work on windows. This is a bug in GHC.
+# --enable-static will pass -staticlib to ghc, which only works for mach-o and elf.
+assert hostPlatform.isWindows -> enableStaticLibraries = false;
```
two equal signs

Ericson2314 commented on pkgs/development/compilers/ghc/head.nix:
```diff
@@ -54,10 +68,11 @@ let
     EXTRA_CC_OPTS += -std=gnu99
   '';
 
+
```
blank line wasn't added in 8.4.1. In general it's nice to diff the different GHCs so they don't unnecessarily drift apart. We really should deduplicate soon...

Ericson2314 commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -268,6 +278,28 @@ stdenv.mkDerivation ({
     for f in "$packageConfDir/"*.conf; do
       sed -i "s,dynamic-library-dirs: .*,dynamic-library-dirs: $dynamicLinksDir," $f
     done
+  '') + (optionalString (isCross && setupHaskellDepends != []) ''
```
Can you abstract this part out like https://github.com/NixOS/nixpkgs/pull/39735/files#diff-d65852670e35e11b4bd12ed1c634abce ? Fine to wait on separate derivations until #39735 to avoid controversy though :)

angerman replied:
I thought about it, but I kept the `Setup.hs` changes out of this diff on purpose.

Ericson2314 commented:
I'm going to rebase onto https://github.com/NixOS/nixpkgs/pull/40442 after I merge it to staging.

Ericson2314 committed:
ghc: Normalize derivations

Ericson2314 committed:
ghc: Handle flavors better

Ericson2314 committed:
ghc 8.4.2, head: Adjust enableShared enableTerminfo for windows

Ericson2314 committed:
haskell generic builder: `setupHaskellDepends` should be `nativeBuildInputs

This is because they are just for Setup.hs, so they are just used at build time
and completely isolated from the normal components' dependencies.

This was previous implemented in 8a8f0408cd9b7fdda1095718107c800057658c44, but
reverted in e69c7f56419589c0d3296e81f47032fa813cca4b because it broken
setup-depends non-cross in haskell shell environments (custom Setup.hs in cross
shell environments has never worked). This version adds a special native
exception to avoid that breakage.

Ericson2314 committed:
Merge branch 'staging' into feature/clean-ghc

Ericson2314 commented:
@peti I wanted to land this somewhere pronto so as to enable more work, but if you could also merge this into master via your normal `haskell-updates` workflow, that would be much appreciated.

peti commented:
It's not that easy, unfortunately, because the patch does not apply to current versions of `master`.
